ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java

```diff
@@ -364,78 +383,100 @@ public void testFindUserByEmployeeId() throws Exception {
         assumeConfigured();
+        assumeGroupsConfigured();
 
         final User user = getUser();
 
         this.dao.save(user);
 
-        List<String[]> paths = Arrays.asList(
-                new String[] {"GoogleApps", "Cru", "Cru"},
-                new String[] {"GoogleApps", "Cru", "AIA"}
-        );
-        String name = "Mail";
-
-        List<Group> groups = Lists.newArrayList();
-        for(String[] path : paths) {
-            Group group = new Group(path, name);
-            groups.add(group);
-            this.dao.addToGroup(user, group);
+        // test adding this user to group1
+        {
+            this.dao.addToGroup(user, group1);
+            final User foundUser = this.dao.findByEmail(user.getEmail(), false);
+            assertEquals(ImmutableSet.of(group1), foundUser.getGroups());
         }
 
-        final User foundUser = this.dao.findByEmail(user.getEmail(), false);
-
-        final Set<Group> foundUserGroups = Sets.newHashSet(foundUser.getGroups());
-
-        Collection<Group> empty = Sets.newHashSet();
-        foundUser.setGroups(empty);
-        Assert.assertEquals(user, foundUser);
+        // test adding this user to group2
+        {
+            this.dao.addToGroup(user, group2);
+            final User foundUser = this.dao.findByEmail(user.getEmail(), false);
+            assertEquals(ImmutableSet.of(group1, group2), foundUser.getGroups());
+        }
 
-        Assert.assertTrue(foundUserGroups.size() == groups.size());
+        // test removing this user from group 1
+        {
+            this.dao.removeFromGroup(user, group1);
+            final User foundUser = this.dao.findByEmail(user.getEmail(), false);
+            assertEquals(ImmutableSet.of(group2), foundUser.getGroups());
+        }
 
-        for(Group foundGroup : foundUserGroups) {
-            Boolean match = false;
-            for(Group group : groups) {
-                if(group.equals(foundGroup)) {
-                    match = true;
-                    break;
-                }
-            }
-            Assert.assertTrue(match);
+        // test removing this user from group 2
+        {
+            this.dao.removeFromGroup(user, group2);
+            final User foundUser = this.dao.findByEmail(user.getEmail(), false);
+            assertEquals(ImmutableSet.of(), foundUser.getGroups());
         }
     }
 
     @Test
-    public void testRemoveFromGroup() throws Exception {
+    public void testFindAllByGroup() throws Exception {
         assumeConfigured();
+        assumeGroupsConfigured();
 
-        final User user = getUser();
+        // create a couple users for testing
+        final User user1 = getUser();
+        final User user2 = getUser();
+        user2.setDeactivated(true);
+        this.dao.save(user1);
+        this.dao.save(user2);
 
-        this.dao.save(user);
+        final Function<User, String> FUNCTION_GUID = new Function<User, String>() {
```

Comment:
My two cents.

There always seems to be tradeoff between providing certain potentially valuable (but currently unused) functionality (e.g. `IllegalGroupDnException`) and/or 'creeping elegance' in the code vs. leaving it undeveloped until requirements warrant it. I don't have a hard opinion either way. I would simply say that if you've felt the need to develop the functionality or add the improvement, then it's probably worth providing it for potential future use.

So, I would say add `Function` `FUNCTION_GUID` to the `User` object, the only caveat being that there be clear delineation between types of methods, that is, getter/setters vs not meant for public use (e.g. `public <T> T getImplMeta`), and this helper method. Whether you distinguish by way of comments and method placement in the class, or whether you provide another 'helper' class, I leave to you, though I would lean towards the former, as you originally suggested.